### PR TITLE
Add veMesh simple strategy

### PR DESCRIPTION
**Pull request by AlehNat**

### CODE REVIEW

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+  // ------------------ GOV actions --------------------------
+
+  /// @dev Manual withdraw for any emergency purposes
+  function manualWithdraw() external restricted {
+    IERC20(_underlying()).safeTransfer(_vault(), IERC20(_underlying()).balanceOf(address(this)));
+  }
+
+  function addVoting(address exchange, uint256 amount) external restricted {
+    require(exchange != address(0), "Exchange address should be specified");
```
Comment (bogdoslavik) [inline contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:69]: use custom errors https://blog.soliditylang.org/2021/04/21/custom-errors/

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    poolVoting.addVoting(exchange, amount);
+    emit VotingAdded(exchange, amount);
+  }
+
+  function removeVoting(address exchange, uint256 amount) external restricted {
+    require(exchange != address(0), "Exchange address should be specified");
```
Comment (bogdoslavik) [inline contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:75]: use custom errors

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      amount = amount / _MESH_PRECISION; // rounding
+      IERC20(_underlying()).safeApprove(address(vMesh), 0);
+      IERC20(_underlying()).safeApprove(address(vMesh), amount * _MESH_PRECISION);
+      vMesh.lockMESH(amount, _MAX_LOCK_PERIOD);
+    }
+  }
+
+  /// @dev Not supported by MESH
+  function withdrawAndClaimFromPool(uint256) internal pure override {
+    revert("MSS: Withdraw forbidden");
```
Comment (bogdoslavik) [inline contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:111]: use custom errors

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  /// @dev Not supported by MESH
+  function emergencyWithdrawFromPool() internal pure override {
+    revert("MSS: Method not supported");
```
Comment (bogdoslavik) [inline contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:116]: use custom errors

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  /// @dev In this version MESH rewards are not liquidated.
+  function liquidateReward() internal override {}
+
+  /// @dev No claimable tokens
+  function readyToClaim() external view override returns (uint256[] memory) {
```
Comment (bogdoslavik) [inline contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:123]: you can name output variable and do not redeclare it
```
returns (uint256[] memory toClaim ) {
  toClaim = new uint256[](_rewardTokens.length);
}
```

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  /// @dev Return full amount of staked tokens
+  function poolTotalAmount() external view override returns (uint256) {
+    return IERC20(_underlying()).balanceOf(address(vMesh));
+  }
+
+  /// @dev Platform name for statistical purposes
+  /// @return Platform enum index
+  function platform() external override pure returns (Platform) {
```
Comment (bogdoslavik) [inline contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:135]: Looks like possible to declare it one line as constant:
Platform public constant override platform = Platform.MESH;

File: test/strategies/matic/mesh/MeshStakeStrategyTest.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import {DeployInfo} from "../../DeployInfo";
+import {SpecificStrategyTest} from "../../SpecificStrategyTest";
+import {SignerWithAddress} from "@nomiclabs/hardhat-ethers/signers";
+import {CoreContractsWrapper} from "../../../CoreContractsWrapper";
+import {DeployerUtils} from "../../../../scripts/deploy/DeployerUtils";
+import {IStrategy, SmartVault, SmartVault__factory, StrategyMeshStaking} from "../../../../typechain";
+import {ToolsContractsWrapper} from "../../../ToolsContractsWrapper";
+import {universalStrategyTest} from "../../UniversalStrategyTest";
+import {MeshStakingDoHardWork} from "./MeshStakingDoHardWork";
+import {VeMeshSpecificTests} from "./veMeshSpecificTests";
```
Comment (bogdoslavik) [inline test/strategies/matic/mesh/MeshStakeStrategyTest.ts:15]: **V**eMeshSpecificTests.ts - first uppercase

**Review by bogdoslavik — COMMENTED** [review]
Should pass all tests

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  function doHardWork() external onlyNotPausedInvesting override hardWorkers {
+    vMesh.claimReward();
+  }
+
+  /// @dev Stake Mesh to vMesh
+  function depositToPool(uint256 amount) internal override {
+    if (amount > _MESH_PRECISION) {
+      // lock on max period
+      // mesh allows only integer values w/o precision e.g 1 mesh token
+      amount = amount / _MESH_PRECISION; // rounding
```
Comment (belbix) [inline contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:102]: Are you sure? Could you ask them?

File: test/strategies/UniversalStrategyTest.ts [hunk 9]
```diff
@@ -123,24 +123,24 @@ async function universalStrategyTest(
...
-        vault,
-        strategy,
-        balanceTolerance,
-      ).start(userBalance, loops, loopValue, advanceBlocks);
-    });
-
-    it("common test should be ok", async () => {
-      await StrategyTestUtils.commonTests(strategy, underlying);
-    });
+    // it("doHardWork loop", async function () {
```
Comment (belbix) [inline test/strategies/UniversalStrategyTest.ts:126]: ??? Uncoment please

**Review by belbix — CHANGES_REQUESTED** [review]

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+  // --------------------- CONSTANTS -------------------------------
+  /// @notice Strategy type for statistical purposes
+  string public constant override STRATEGY_NAME = "MeshStakingStrategyBase";
+  /// @notice Version of the contract
+  /// @dev Should be incremented when contract changed
+  string public constant VERSION = "1.0.0";
+  /// @dev 10% buybacks, 90% of vested Mesh should go to the vault rewards (not autocompound)
+  uint256 private constant _BUY_BACK_RATIO = 10_00;
+  uint256 private constant _MAX_LOCK_PERIOD = 1555200000;
```
[the remark this answers is not in the document]
Comment (AlehNat) [reply contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:34]: It is custom value provided by Mesh team like 
86400sec * 360 days * 50 years = 155200000

File: contracts/base/strategies/mesh/MeshStakingStrategyBase.sol (hunk 5 above)
Comment (AlehNat) [reply contracts/base/strategies/mesh/MeshStakingStrategyBase.sol:123]: Done

**Pull request closed without merging**